checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -141,6 +327,123 @@ private void visitAccess(ExpressionTree indexTree, ExpressionTree arrTree) {
+
+        if (elt != null
+                && (elt.getKind() == ElementKind.FIELD
+                        || elt.getKind() == ElementKind.PARAMETER
+                        || elt.getKind() == ElementKind.LOCAL_VARIABLE)
+                && !ElementUtils.isEffectivelyFinal(elt)) {
+            TypeKind typeKind = InternalUtils.typeOf(varTree).getKind();
+            Receiver field = FlowExpressions.internalReprOf(atypeFactory, (ExpressionTree) varTree);
+            ReassignmentKind kind;
+            if (elt.getKind() == ElementKind.FIELD) {
+                if (typeKind == TypeKind.ARRAY) {
+                    kind = ReassignmentKind.ARRAY_FIELD_REASSIGNMENT;
+                } else if (!typeKind.isPrimitive()) {
+                    kind = UpperBoundUtil.ReassignmentKind.NON_ARRAY_FIELD_REASSIGNMENT;
+                } else {
+                    return;
+                }
+            } else if (typeKind == TypeKind.ARRAY) {
+                kind = UpperBoundUtil.ReassignmentKind.LOCAL_VAR_REASSIGNMENT;
+            } else {
+                return;
+            }
+            checkAnnotationsInClass(elt, field, varTree, getCurrentPath(), kind);
+        }
+    }
+
+    /**
+     * Checks whether any annotations in scope are dependent and need to be invalidated by the given
+     * side effect.
+     *
+     * @param elt An element corresponding to the side effect that caused the invalidation.
+     * @param possiblyInvalidatedRec A receiver representing the possibly invalidated expression.
+     * @param tree The tree on which the side effect occurs. Used in error reporting.
+     * @param path The path on which the side effect occurs. Used for parsing.
+     * @param reassignmentKind The type of side effect. See {@link UpperBoundUtil.ReassignmentKind}.
+     */
+    private void checkAnnotationsInClass(
+            Element elt,
+            Receiver possiblyInvalidatedRec,
+            Tree tree,
+            TreePath path,
+            UpperBoundUtil.ReassignmentKind reassignmentKind) {
+        List<? extends Element> enclosedElts =
+                enclosingElementsCache.containsKey(elt) ? enclosingElementsCache.get(elt) : null;
+        if (enclosedElts == null) {
+            enclosedElts = ElementUtils.enclosingClass(elt).getEnclosedElements();
+            enclosingElementsCache.put(elt, enclosedElts);
+        }
+        List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(enclosedElts);
+        for (AnnotatedTypeMirror atm : enclosedTypes) {
+            List<Receiver> rs =
+                    UpperBoundUtil.getDependentReceivers(atm.getAnnotations(), path, atypeFactory);
+            for (Receiver r : rs) {
+                ReassignmentError result =
+                        UpperBoundUtil.isSideEffected(r, possiblyInvalidatedRec, reassignmentKind);
+                if (result != ReassignmentError.NO_ERROR) {
+                    checker.report(Result.failure(result.errorKey, atm), tree);
```

Comment:
The error keys have require different number of arguments, so this causes errors such as:
```
 Invalid format string: "[reassignment.not.permitted] Reassigning the variable %s is not permitted, because the annotation %s depends on it. Reassigning %s would invalidate that annotation." args: [@LTLengthOf(value="array", offset={}) int]
```